rfc/2023/2023-05-01-va-rfc-frontend-performance.md

```diff
@@ -0,0 +1,172 @@
+
+### Address performance issues that affect all applications
+
+Sitewide performance issues can be addressed to improve performance across VA.gov applications. The **shared header and footer** are React applications that are part of the `static-pages.entry.js` bundle, which is 1.397MB as of April 2023. The header and footer should be refactored to remove React and use build-time generated HTML, with lightweight web components to add interactivity and manage accessibility. Improving performance of the header and footer will have a positive impact across VA.gov and subdomains that inject the header and footer. This includes teamsites, which load the header and footer asynchronously.
+
+### Raise visibility of performance data by enhancing existing reports
+
+Another step is to update the [Frontend Support Dashboard - Performance Report](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/frontend-support-dashboard/lighthouse-performance-report/) to include all applications. This will raise visibility of performance issues, both per application and sitewide. This step will also help teams remember to review those reports periodically, this will also help teams understand where their applications experience performance issues.
+
+### Have VFS teams set file size limits tailored to their applications
+
+["The Performance Inequality Gap," 2023](https://infrequently.org/2022/12/performance-baseline-2023/) suggests file size limits for HTML, CSS, and JavaScript, for example "~151KiB of HTML/CSS/fonts and ~300-350KiB of JavaScript (gzipped)". Using that report as a guide, we can determine reasonable file size limits and include them in our own performance budgets. VFS teams would use these baselines, along with the information about their application's real world performance provided by Datadog RUM and the Performance Report dashboard, to manage performance budgets for their applications.
+
+### Leverage existing tools to reduce code duplcation
+
+As mentioned above, we can leverage our existing build tool, Webpack, to [prevent duplication](https://webpack.js.org/guides/code-splitting/#prevent-duplication) of code, removing the multiple copies of the same libraries in bundles. Webpack also provides options for [notifying teams of performance issues](https://webpack.js.org/configuration/performance/), another way to help teams understand performance and adopt a budget.
+
+## Risks
+
+
+>List the risks of this approach
+> * There are always risks. What are the risks of this solution?
+> * These are the things people will bring up in opposition to your idea or plans. Acknowledge them.
+
+- There are large assets on the page such as Google Analytics and Digital Voice scripts that we cannot optimize or remove. Performance budgets can and should acknowledge this.
```

Comment:
Why can't we ask the Digital Voice team to optimize?

Reply:
I think we could! The "we" in that sentence was VFS/Platform teams. From looking at those scripts, I _think_ Digital Voice code is (primarily?) Medallia Analytics… so that looks like a different set of conversations than what can VFS/Platform teams do. Good point, however!